packages/perspective-viewer-d3fc/src/js/d3fc/extent/extent.js

```diff
@@ -0,0 +1,137 @@
+
+                for (let i = 0; i < data.length; i++) {
+                    let value = accessor(data[i], i);
+                    if (Array.isArray(value)) {
+                        values.push.apply(values, toConsumableArray(value));
+                    } else {
+                        values.push(value);
+                    }
+                }
+            }
+        } catch (err) {
+            _didIteratorError = true;
+            _iteratorError = err;
+        } finally {
+            try {
+                if (!_iteratorNormalCompletion && _iterator.return) {
+                    _iterator.return();
+                }
+            } finally {
+                if (_didIteratorError) {
+                    throw _iteratorError;
+                }
+            }
+        }
+
+        let extent$$1 = [d3Array.min(values), d3Array.max(values)];
+
+        extent$$1[0] = extent$$1[0] == null ? d3Array.min(include) : d3Array.min([extent$$1[0]].concat(toConsumableArray(include)));
+        extent$$1[1] = extent$$1[1] == null ? d3Array.max(include) : d3Array.max([extent$$1[1]].concat(toConsumableArray(include)));
+
+        if (symmetricalAbout != null) {
+            let halfRange = Math.max(Math.abs(extent$$1[1] - symmetricalAbout), Math.abs(extent$$1[0] - symmetricalAbout));
+            extent$$1[0] = symmetricalAbout - halfRange;
+            extent$$1[1] = symmetricalAbout + halfRange;
+        }
+
+        paddingStrategy(extent$$1);
+
+        return extent$$1;
+    };
+
+    instance.accessors = function() {
+        if (!arguments.length) {
+            return accessors;
+        }
+        accessors = arguments.length <= 0 ? undefined : arguments[0];
+        return instance;
+    };
+
+    //This function points directly at the paddingStrategy child object's properties for backwards-compatibility. DEPRECATED.
+    instance.pad = function() {
+        if (!arguments.length) {
+            return paddingStrategy.pad;
+        }
+        paddingStrategy.pad(arguments.length <= 0 ? undefined : arguments[0]);
+        return instance;
+    };
+
+    //This function points directly at the paddingStrategy child object's properties for backwards-compatibility. DEPRECATED.
+    instance.padUnit = function() {
+        if (!arguments.length) {
+            return paddingStrategy.padUnit;
+        }
+        paddingStrategy.padUnit(arguments.length <= 0 ? undefined : arguments[0]);
+        return instance;
+    };
+
+    instance.include = function() {
+        if (!arguments.length) {
+            return include;
+        }
+        include = arguments.length <= 0 ? undefined : arguments[0];
+        return instance;
+    };
+
+    instance.symmetricalAbout = function() {
+        if (!arguments.length) {
+            return symmetricalAbout;
+        }
+        symmetricalAbout = arguments.length <= 0 ? undefined : arguments[0];
+        return instance;
+    };
+
+    instance.paddingStrategy = function() {
```

Comment:
The approach looks good, and non-breaking.

This might be something to discuss on the d3fc issue, but I'm wondering whether we could overload the `pad()` function instead of adding a new `paddingStrategy` function.

i.e. When the value is `set`, it would need to check whether it was given an array (call through to the default padding trategy) or else a function (replace the padding strategy). When the value 'getter' is called, it would need to know whether it's still using the default strategy and return the current array (for backwards compatibility), and otherwise return the function. That might be a bit too fiddly for their tastes, so needs discussing.

Reply:
Interesting idea. I quite like it. I can't find a way to implement it though without adding a property that I suspect will cause issues in future. The problem is distinguishing between the defaultPadding and other strategies. 

For example if there's an additional property on the paddingStrats which defines whether they're _backwardCompatible_ then if someone creates a new strategy I can imagine them overlooking the necessary override and getting strange results. 

Or if a flag is set when a new strategy is loaded, is there a way to identify that they've not just loaded in the defaultStrategy explicitly? Although, if they're doing that, it would require a change anyway, and wouldn't break existing implementations, so perhaps that would be fine.

Ultimately I would sooner not make this change, but I'm not married to the idea of shirking this if we think it's truly the better strategy. As you say, the d3fc team may have their own views on this, so I don't think any word here will be final.

Reply:
I would say that if someone calls `.pad(someFunction)`, then when they call `.pad()` they should expect to get `someFunction` back, whether or not it was the default strategy.

I agree that this might be controversial, but I think it's a topic worth discussing over at the d3fc issue.